File: backend/app/schemas/bao_cao_xep_loai.py
"""
app/schemas/bao_cao_xep_loai.py
===============================
Pydantic schemas cho Báo cáo Xếp loại Chất lượng Công chức.

Bao gồm schemas cho:
- BaoCaoXepLoaiResponse: Response chi tiết báo cáo
- ChiTietXepLoaiResponse: Response chi tiết từng CC
- DeXuatXepLoaiRequest: Request điều chỉnh xếp loại (Đội trưởng)
- QuyetDinhXepLoaiRequest: Request điều chỉnh xếp loại (CCT)
- PheDuyetBaoCaoRequest: Request phê duyệt/từ chối báo cáo

Phiên bản: 1.0 (28/01/2026)
"""

from datetime import datetime
from decimal import Decimal
from typing import Optional, List
from uuid import UUID
from enum import Enum

from pydantic import BaseModel, Field, ConfigDict, field_validator


# =============================================================================
# ENUMS
# =============================================================================

class TrangThaiBaoCaoEnum(str, Enum):
    """Trạng thái báo cáo."""
    NHAP = "NHAP"
    CHO_PHE_DUYET = "CHO_PHE_DUYET"
    DA_PHE_DUYET = "DA_PHE_DUYET"
    TU_CHOI = "TU_CHOI"


class XepLoaiEnum(str, Enum):
    """Mức xếp loại chất lượng."""
    A = "A"
    B = "B"
    C = "C"
    D = "D"
    E = "E"


# =============================================================================
# NESTED SCHEMAS
# =============================================================================

class DonViBrief(BaseModel):
    """Thông tin tóm tắt đơn vị."""
    model_config = ConfigDict(from_attributes=True)
    
    id: UUID
    ma_don_vi: str
    ten_don_vi: str


class CongChucBrief(BaseModel):
    """Thông tin tóm tắt công chức."""
    model_config = ConfigDict(from_attributes=True)
    
    id: UUID
    ma_cc: str
    ho_ten: str
    chuc_vu: Optional[str] = None


class ThongKeBaoCao(BaseModel):
    """Thống kê xếp loại trong báo cáo."""
    tong_cong_chuc: int = 0
    so_loai_a: int = 0
    so_loai_b: int = 0
    so_loai_c: int = 0
    so_loai_d: int = 0
    so_loai_e: int = 0


# =============================================================================
# CHI TIẾT XẾP LOẠI
# =============================================================================

class ChiTietXepLoaiResponse(BaseModel):
    """Response chi tiết xếp loại của 1 công chức."""
    model_config = ConfigDict(from_attributes=True)
    
    id: UUID
    cong_chuc_id: UUID
    cong_chuc: Optional[CongChucBrief] = None
    
    # Phân loại
    is_lanh_dao: bool = False
    
    # Điểm từ hệ thống
    diem_tieu_chi_chung: Decimal = Field(default=Decimal("0"), description="Điểm tiêu chí chung (0-30)")
    diem_kpi: Decimal = Field(default=Decimal("0"), description="Điểm KPI (0-70)")
    diem_tong: Decimal = Field(default=Decimal("0"), description="Điểm tổng (0-100)")
    xep_loai_he_thong: str = Field(description="Xếp loại hệ thống tự tính")
    
    # Đề xuất của Đội trưởng
    xep_loai_de_xuat: Optional[str] = None
    ly_do_dieu_chinh_dt: Optional[str] = None
    
    # Quyết định của CCT
    xep_loai_quyet_dinh: Optional[str] = None
    ly_do_dieu_chinh_cct: Optional[str] = None
    
    # Trạng thái từ chối
    bi_tu_choi: bool = False
    ly_do_tu_choi: Optional[str] = None
    
    # Ghi chú
    ghi_chu: Optional[str] = None


class DeXuatXepLoaiRequest(BaseModel):
    """
    Request điều chỉnh xếp loại đề xuất (Đội trưởng).
    
    Validation:
    - Nếu xep_loai_de_xuat != xep_loai_he_thong → ly_do_dieu_chinh BẮT BUỘC
    """
    xep_loai_de_xuat: XepLoaiEnum = Field(
        ...,
        description="Mức xếp loại đề xuất (A/B/C/D/E)"
    )
    ly_do_dieu_chinh: Optional[str] = Field(
        default=None,
        max_length=2000,
        description="Lý do điều chỉnh (bắt buộc nếu khác hệ thống)"
    )
    ghi_chu: Optional[str] = Field(
        default=None,
        max_length=1000,
        description="Ghi chú bổ sung"
    )


class QuyetDinhXepLoaiRequest(BaseModel):
    """
    Request điều chỉnh xếp loại quyết định (CCT).
    
    Validation:
    - Nếu xep_loai_quyet_dinh != xep_loai_de_xuat → ly_do_dieu_chinh BẮT BUỘC
    """
    xep_loai_quyet_dinh: XepLoaiEnum = Field(
        ...,
        description="Mức xếp loại quyết định (A/B/C/D/E)"
    )
    ly_do_dieu_chinh: Optional[str] = Field(
        default=None,
        max_length=2000,
        description="Lý do điều chỉnh (bắt buộc nếu khác đề xuất)"
    )


# =============================================================================
# BÁO CÁO XẾP LOẠI
# =============================================================================

class BaoCaoXepLoaiResponse(BaseModel):
    """Response chi tiết báo cáo xếp loại."""
    model_config = ConfigDict(from_attributes=True)
    
    id: UUID
    
    # Thông tin báo cáo
    don_vi_id: UUID
    don_vi: Optional[DonViBrief] = None
    thang: int
    nam: int
    
    # Người lập
    nguoi_lap_id: UUID
    nguoi_lap: Optional[CongChucBrief] = None
    ngay_lap: Optional[datetime] = None
    
    # Trạng thái
    trang_thai: str
    trang_thai_ten: str = Field(default="", description="Tên tiếng Việt của trạng thái")
    
    # Phê duyệt
    nguoi_phe_duyet_id: Optional[UUID] = None
    nguoi_phe_duyet: Optional[CongChucBrief] = None
    ngay_phe_duyet: Optional[datetime] = None
    y_kien_phe_duyet: Optional[str] = None
    
    # Thống kê
    tong_cong_chuc: int = 0
    so_loai_a: int = 0
    so_loai_b: int = 0
    so_loai_c: int = 0
    so_loai_d: int = 0
    so_loai_e: int = 0
    canh_bao_ty_le_a: bool = False
    
    # Chi tiết
    chi_tiet: List[ChiTietXepLoaiResponse] = []
    
    # Timestamps
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class BaoCaoXepLoaiBriefResponse(BaseModel):
    """Response tóm tắt báo cáo (cho danh sách)."""
    model_config = ConfigDict(from_attributes=True)
    
    id: UUID
    
    # Thông tin báo cáo
    don_vi: Optional[DonViBrief] = None
    thang: int
    nam: int
    
    # Người lập
    nguoi_lap: Optional[CongChucBrief] = None
    ngay_lap: Optional[datetime] = None
    
    # Trạng thái
    trang_thai: str
    
    # Thống kê
    tong_cong_chuc: int = 0
    so_loai_a: int = 0
    so_loai_b: int = 0
    so_loai_c: int = 0
    so_loai_d: int = 0
    so_loai_e: int = 0
    canh_bao_ty_le_a: bool = False


# =============================================================================
# GỬI DUYỆT & PHÊ DUYỆT
# =============================================================================

class GuiDuyetBaoCaoResponse(BaseModel):
    """Response sau khi gửi duyệt báo cáo."""
    id: UUID
    trang_thai: str
    ngay_lap: datetime
    canh_bao_ty_le_a: bool = False
    thong_ke: ThongKeBaoCao


class ChiTietTuChoiItem(BaseModel):
    """Chi tiết từ chối 1 CC cụ thể."""
    chi_tiet_id: UUID = Field(..., description="ID chi tiết xếp loại")
    ly_do: str = Field(
        ...,
        min_length=1,
        max_length=1000,
        description="Lý do từ chối"
    )


class PheDuyetBaoCaoRequest(BaseModel):
    """
    Request phê duyệt/từ chối báo cáo (CCT).
    
    Logic:
    - APPROVE: Phê duyệt toàn bộ báo cáo
    - REJECT: Từ chối, có thể chỉ định danh sách CC bị từ chối
    """
    action: str = Field(
        ...,
        description="APPROVE hoặc REJECT"
    )
    y_kien: Optional[str] = Field(
        default=None,
        max_length=2000,
        description="Ý kiến của CCT"
    )
    chi_tiet_tu_choi: Optional[List[ChiTietTuChoiItem]] = Field(
        default=None,
        description="Danh sách CC bị từ chối (chỉ khi REJECT)"
    )
    
    @field_validator("action")
    @classmethod
    def validate_action(cls, v: str) -> str:
        if v not in ["APPROVE", "REJECT"]:
            raise ValueError("action phải là APPROVE hoặc REJECT")
        return v


class PheDuyetBaoCaoResponse(BaseModel):
    """Response sau khi phê duyệt/từ chối báo cáo."""
    id: UUID
    trang_thai: str
    ngay_phe_duyet: Optional[datetime] = None
    so_cc_bi_tu_choi: int = 0


# =============================================================================
# THỐNG KÊ
# =============================================================================

class ThongKeXepLoaiDonVi(BaseModel):
    """Thống kê xếp loại của 1 đơn vị."""
    don_vi: DonViBrief
    tong: int = 0
    A: int = 0
    B: int = 0
    C: int = 0
    D: int = 0
    E: int = 0
    trang_thai: str = "CHUA_CO"  # Trạng thái báo cáo


class ThongKeXepLoaiChiCuc(BaseModel):
    """Thống kê xếp loại toàn Chi cục."""
    thang: int
    nam: int
    tong_cong_chuc: int = 0
    theo_xep_loai: dict = Field(
        default_factory=lambda: {"A": 0, "B": 0, "C": 0, "D": 0, "E": 0}
    )
    theo_don_vi: List[ThongKeXepLoaiDonVi] = []


# =============================================================================
# HELPER FUNCTIONS
# =============================================================================

TRANG_THAI_TEN = {
    "NHAP": "Đang soạn",
    "CHO_PHE_DUYET": "Chờ phê duyệt",
    "DA_PHE_DUYET": "Đã phê duyệt",
    "TU_CHOI": "Bị từ chối",
}


def get_trang_thai_ten(trang_thai: str) -> str:
    """Lấy tên tiếng Việt của trạng thái."""
    return TRANG_THAI_TEN.get(trang_thai, trang_thai)

File: frontend/src/types/bao-cao-xep-loai.ts
/**
 * src/types/bao-cao-xep-loai.ts
 * ==============================
 * TypeScript interfaces cho Module Báo cáo Xếp loại Chất lượng Công chức.
 * 
 * Theo Điều 17 - Quy chế đánh giá KPI Chi cục Hải quan KV8.
 * 
 * Version: 1.0.0 (28/01/2026)
 */

// =============================================================================
// ENUMS
// =============================================================================

/**
 * Trạng thái báo cáo xếp loại.
 */
export enum TrangThaiBaoCao {
  NHAP = 'NHAP',
  CHO_PHE_DUYET = 'CHO_PHE_DUYET',
  DA_PHE_DUYET = 'DA_PHE_DUYET',
  TU_CHOI = 'TU_CHOI',
}

/**
 * Xếp loại chất lượng công chức.
 */
export type XepLoaiChatLuong = 'A' | 'B' | 'C' | 'D' | 'E';

// =============================================================================
// NESTED INTERFACES
// =============================================================================

/**
 * Thông tin đơn vị tóm tắt.
 */
export interface IDonViBrief {
  id: string;
  ma_don_vi: string;
  ten_don_vi: string;
}

/**
 * Thông tin công chức tóm tắt.
 */
export interface ICongChucBrief {
  id: string;
  ma_cc: string;
  ho_ten: string;
  chuc_vu?: string;
}

// =============================================================================
// TIÊU CHÍ CHUNG — điểm "Đánh giá tháng" (override của LĐ/CCT)
// Dùng chung cho modal SuaDiemTieuChiModal + trang Điều chỉnh tiêu chí chung.
// =============================================================================

export interface ITieuChiChungItemDgt {
  ma_tieu_chi: string;
  ten_tieu_chi: string;
  nhom_tieu_chi?: number;
  diem_toi_da: number;
  diem_tu_cham?: number | null;
  diem_phe_duyet?: number | null;
  // Điểm "Đánh giá tháng" — LĐ/CCT chỉnh ở giai đoạn báo cáo xếp loại.
  diem_danh_gia_thang?: number | null;
  diem?: number;
  trang_thai?: string;
}

export interface ITieuChiChungBrief {
  danh_gia_thang_id?: string | null;
  trang_thai?: string;
  tieu_chi?: ITieuChiChungItemDgt[];
}

// =============================================================================
// CHI TIẾT XẾP LOẠI
// =============================================================================

/**
 * Chi tiết xếp loại 1 công chức trong báo cáo.
 */
export interface IChiTietXepLoai {
  id: string;
  cong_chuc_id: string;
  cong_chuc: ICongChucBrief;
  is_lanh_dao: boolean;
  
  // Điểm từ hệ thống (snapshot)
  diem_tieu_chi_chung: number;
  diem_kpi: number;
  diem_tong: number;
  xep_loai_he_thong: XepLoaiChatLuong;
  
  // Đề xuất của Đội trưởng
  xep_loai_de_xuat: XepLoaiChatLuong | null;
  ly_do_dieu_chinh_dt: string | null;
  
  // Quyết định của CCT
  xep_loai_quyet_dinh: XepLoaiChatLuong | null;
  ly_do_dieu_chinh_cct: string | null;
  
  // Từ chối
  bi_tu_choi: boolean;
  ly_do_tu_choi: string | null;
  
  // Ghi chú
  ghi_chu: string | null;
}

// =============================================================================
// BÁO CÁO XẾP LOẠI
// =============================================================================

/**
 * Báo cáo xếp loại đầy đủ (với chi tiết).
 */
export interface IBaoCaoXepLoai {
  id: string;
  don_vi_id: string;
  don_vi: IDonViBrief;
  thang: number;
  nam: number;
  
  // Người lập
  nguoi_lap: ICongChucBrief;
  ngay_lap: string | null;
  
  // Trạng thái
  trang_thai: TrangThaiBaoCao;
  trang_thai_ten?: string;
  
  // Phê duyệt
  nguoi_phe_duyet?: ICongChucBrief | null;
  ngay_phe_duyet?: string | null;
  y_kien_phe_duyet?: string | null;
  
  // Thống kê
  tong_cong_chuc: number;
  so_loai_a: number;
  so_loai_b: number;
  so_loai_c: number;
  so_loai_d: number;
  so_loai_e: number;
  
  // Cảnh báo
  canh_bao_ty_le_a: boolean;
  
  // Chi tiết
  chi_tiet: IChiTietXepLoai[];
}

/**
 * Báo cáo tóm tắt (danh sách chờ phê duyệt).
 */
export interface IBaoCaoXepLoaiSummary {
  id: string;
  don_vi: IDonViBrief;
  thang: number;
  nam: number;
  nguoi_lap: ICongChucBrief;
  ngay_lap: string;
  trang_thai: TrangThaiBaoCao;
  tong_cong_chuc: number;
  so_loai_a: number;
  so_loai_b: number;
  so_loai_c: number;
  so_loai_d: number;
  so_loai_e: number;
  canh_bao_ty_le_a: boolean;
}

// =============================================================================
// REQUEST INTERFACES
// =============================================================================

/**
 * Request điều chỉnh xếp loại đề xuất (Đội trưởng).
 */
export interface IDeXuatXepLoaiRequest {
  xep_loai_de_xuat: XepLoaiChatLuong;
  ly_do_dieu_chinh?: string;
  ghi_chu?: string;
}

/**
 * Request điều chỉnh xếp loại quyết định (CCT).
 */
export interface IQuyetDinhXepLoaiRequest {
  xep_loai_quyet_dinh: XepLoaiChatLuong;
  ly_do_dieu_chinh?: string;
}

/**
 * Chi tiết từ chối 1 CC.
 */
export interface IChiTietTuChoi {
  chi_tiet_id: string;
  ly_do: string;
}

/**
 * Request phê duyệt/từ chối báo cáo.
 */
export interface IPheDuyetBaoCaoRequest {
  action: 'APPROVE' | 'REJECT';
  y_kien?: string;
  chi_tiet_tu_choi?: IChiTietTuChoi[];
}

// =============================================================================
// RESPONSE INTERFACES
// =============================================================================

/**
 * Response gửi duyệt.
 */
export interface IGuiDuyetResponse {
  id: string;
  trang_thai: TrangThaiBaoCao;
  ngay_lap: string;
  canh_bao_ty_le_a: boolean;
  thong_ke: {
    tong_cong_chuc: number;
    so_loai_a: number;
    so_loai_b: number;
    so_loai_c: number;
    so_loai_d: number;
    so_loai_e: number;
  };
}

/**
 * Response phê duyệt.
 */
export interface IPheDuyetResponse {
  id: string;
  trang_thai: TrangThaiBaoCao;
  ngay_phe_duyet?: string;
  so_cc_bi_tu_choi?: number;
}

// =============================================================================
// THỐNG KÊ
// =============================================================================

/**
 * Thống kê theo đơn vị.
 */
export interface IThongKeTheoDonVi {
  don_vi: IDonViBrief;
  tong: number;
  A: number;
  B: number;
  C: number;
  D: number;
  E: number;
  trang_thai: TrangThaiBaoCao | null;
}

/**
 * Thống kê toàn Chi cục.
 */
export interface IThongKeXepLoai {
  thang: number;
  nam: number;
  tong_cong_chuc: number;
  theo_xep_loai: {
    A: number;
    B: number;
    C: number;
    D: number;
    E: number;
  };
  theo_don_vi: IThongKeTheoDonVi[];
}

// =============================================================================
// HELPER FUNCTIONS
// =============================================================================

/**
 * Lấy màu cho xếp loại.
 */
export function getXepLoaiColor(xepLoai: XepLoaiChatLuong): {
  bg: string;
  text: string;
  border: string;
} {
  const colors: Record<XepLoaiChatLuong, { bg: string; text: string; border: string }> = {
    'A': { bg: 'bg-green-100', text: 'text-green-700', border: 'border-green-300' },
    'B': { bg: 'bg-blue-100', text: 'text-blue-700', border: 'border-blue-300' },
    'C': { bg: 'bg-yellow-100', text: 'text-yellow-700', border: 'border-yellow-300' },
    'D': { bg: 'bg-orange-100', text: 'text-orange-700', border: 'border-orange-300' },
    'E': { bg: 'bg-red-100', text: 'text-red-700', border: 'border-red-300' },
  };
  return colors[xepLoai];
}

/**
 * Lấy tên trạng thái báo cáo.
 */
export function getTrangThaiTen(trangThai: TrangThaiBaoCao): string {
  const names: Record<TrangThaiBaoCao, string> = {
    [TrangThaiBaoCao.NHAP]: 'Đang soạn',
    [TrangThaiBaoCao.CHO_PHE_DUYET]: 'Chờ phê duyệt',
    [TrangThaiBaoCao.DA_PHE_DUYET]: 'Đã phê duyệt',
    [TrangThaiBaoCao.TU_CHOI]: 'Bị từ chối',
  };
  return names[trangThai];
}

/**
 * Lấy màu trạng thái.
 */
export function getTrangThaiColor(trangThai: TrangThaiBaoCao): string {
  const colors: Record<TrangThaiBaoCao, string> = {
    [TrangThaiBaoCao.NHAP]: 'bg-gray-100 text-gray-700',
    [TrangThaiBaoCao.CHO_PHE_DUYET]: 'bg-yellow-100 text-yellow-700',
    [TrangThaiBaoCao.DA_PHE_DUYET]: 'bg-green-100 text-green-700',
    [TrangThaiBaoCao.TU_CHOI]: 'bg-red-100 text-red-700',
  };
  return colors[trangThai];
}

/**
 * Tính xếp loại từ điểm.
 */
export function tinhXepLoai(diemTong: number): XepLoaiChatLuong {
  if (diemTong === 0) return 'E';
  if (diemTong < 50) return 'D';
  if (diemTong < 70) return 'C';
  if (diemTong < 90) return 'B';
  return 'A';
}